### PR TITLE
partners/milvus: allow creating a vectorstore with sparse embeddings

### PR DESCRIPTION
**Pull request by ohadeytan**
# Description
Milvus (and `pymilvus`) recently added the option to use [sparse vectors](https://milvus.io/docs/sparse_vector.md#Sparse-Vector) with appropriate search methods (e.g., `SPARSE_INVERTED_INDEX`) and embeddings (e.g., `BM25`, `SPLADE`).

This PR allow creating a vector store using langchain's `Milvus` class, setting the matching vector field type to `DataType.SPARSE_FLOAT_VECTOR` and the default index type to `SPARSE_INVERTED_INDEX`.

It is only extending functionality, and backward compatible. 

## Note
I also interested in extending the Milvus class further to support multi vector search (aka hybrid search). Will be happy to discuss that. See [here](https://github.com/langchain-ai/langchain/discussions/19955), [here](https://github.com/langchain-ai/langchain/pull/20375), and [here](https://github.com/langchain-ai/langchain/discussions/22886) similar needs. 

Twitter handle: [@ohadeytan ](https://x.com/OhadEytan)

### CODE REVIEW

**Comment by ohadeytan** [comment]
Hi @baskaryan, @efriis, @ccurme, can anyone give some feedback here or direct me to the right people to discuss that?

**Comment by zc277584121** [comment]
@ohadeytan Thank you very much for your contribution. I found some langchain documents and I think vector_store may only mean "dense vector store".  And sparse, hybrid and other capabilities should be placed in "Retriever".

https://python.langchain.com/v0.2/docs/integrations/retrievers/milvus_hybrid_search/
https://python.langchain.com/v0.2/docs/integrations/retrievers/qdrant-sparse/
https://python.langchain.com/v0.1/docs/integrations/retrievers/qdrant-sparse/

 I am not sure if the evidence is sufficient, this may also be needed more suggestion from @baskaryan @efriis @ccurme . Any suggestion will be appreciated

**Review by efriis — COMMENTED** [review]
I'm ok with the integration on the actual vs. Could you change `embedding_function` to take the union of the two types and detect which is passed in instead of adding a new kwarg?

**Comment by ohadeytan** [comment]
> I'm ok with the integration on the actual vs. Could you change `embedding_function` to take the union of the two types and detect which is passed in instead of adding a new kwarg?

I did that (before [this](https://github.com/langchain-ai/langchain/pull/25284/commits/14aee874896d36a749469163850e7842c46cb7a6) commit) but `lint` was complaining because it's different from the ancestor `VectorStore` class. Should I revert and ignore the warnings?

File: libs/partners/milvus/langchain_milvus/vectorstores/milvus.py [hunk 1]
```diff
@@ -513,9 +522,16 @@ def _create_collection(
                 )
             )
         # Create the vector field, supports binary or float vectors
-        fields.append(
-            FieldSchema(self._vector_field, infer_dtype_bydata(embeddings[0]), dim=dim)
-        )
+        if isinstance(self.sparse_embedding_func, BaseSparseEmbedding):
+            fields.append(
+                FieldSchema(self._vector_field, DataType.SPARSE_FLOAT_VECTOR, dim=dim)
```
Comment (zc277584121) [inline libs/partners/milvus/langchain_milvus/vectorstores/milvus.py:527]: FieldSchema(self._vector_field, DataType.SPARSE_FLOAT_VECTOR)
if we pass in arg `dim`, there will raise an error :"dim should not be specified for sparse vector field vector(0)"

File: libs/partners/milvus/tests/integration_tests/vectorstores/test_milvus.py [hunk 2]
```diff
@@ -303,6 +304,32 @@ def test_milvus_enable_dynamic_field_with_partition_key() -> None:
...
+        sparse_embedding_function=sparse_embedding_func,
+        connection_args={"uri": "./milvus_demo.db"},
+        drop_old=True,
+    )
+
+    output = docsearch.similarity_search("Pilgrim", k=1)
+    assert "Pilgrim" in output[0].page_content
+
+
 # if __name__ == "__main__":
```
Comment (zc277584121) [inline libs/partners/milvus/tests/integration_tests/vectorstores/test_milvus.py:333]: could you please add "#     test_milvus_sparse_embeddings()" to main(), although it is a comment. Just for completeness and suggestion

**Comment by zc277584121** [comment]
> I'm ok with the integration on the actual vs. Could you change `embedding_function` to take the union of the two types and detect which is passed in instead of adding a new kwarg?

I agree with that ,if both dense and sparse embeddings are regarded as the same kind of abstraction of embedding.

> I did that (before [this](https://github.com/langchain-ai/langchain/pull/25284/commits/14aee874896d36a749469163850e7842c46cb7a6) commit) but lint was complaining because it's different from the ancestor VectorStore class. Should I revert and ignore the warnings?

I am also concern that lint issue, which may indicate the VectorStore means only dense embedding

**Comment by efriis** [comment]
I'm ok with `type: ignore` on that param for here

File: libs/partners/milvus/tests/integration_tests/vectorstores/test_milvus.py (hunk 2 above)
Comment (ohadeytan) [reply libs/partners/milvus/tests/integration_tests/vectorstores/test_milvus.py]: Done

File: libs/partners/milvus/langchain_milvus/vectorstores/milvus.py (hunk 1 above)
Comment (ohadeytan) [reply libs/partners/milvus/langchain_milvus/vectorstores/milvus.py]: I did not see this error, but removed it anyway, as it's unneeded.

**Comment by ohadeytan** [comment]
> I'm ok with `type: ignore` on that param for here

Looks good to me. I fixed the test and addressed the comments from @zc277584121

File: libs/partners/milvus/langchain_milvus/vectorstores/milvus.py [hunk 3]
```diff
@@ -814,7 +839,7 @@ def add_texts(
 
     def _collection_search(
         self,
-        embedding: List[float],
+        embedding: List[float] | Dict[int, float],
```
Comment (zc277584121) [inline libs/partners/milvus/langchain_milvus/vectorstores/milvus.py:842]: wil be appreciate if change the corresponding Args docstr below

File: libs/partners/milvus/langchain_milvus/vectorstores/milvus.py [hunk 4]
```diff
@@ -975,7 +1000,7 @@ def similarity_search_with_score(
 
     def similarity_search_with_score_by_vector(
         self,
-        embedding: List[float],
+        embedding: List[float] | Dict[int, float],
```
Comment (zc277584121) [inline libs/partners/milvus/langchain_milvus/vectorstores/milvus.py:1003]: ditto

File: libs/partners/milvus/langchain_milvus/vectorstores/milvus.py [hunk 5]
```diff
@@ -1067,7 +1092,7 @@ def max_marginal_relevance_search(
 
     def max_marginal_relevance_search_by_vector(
         self,
-        embedding: list[float],
+        embedding: list[float] | dict[int, float],
```
Comment (zc277584121) [inline libs/partners/milvus/langchain_milvus/vectorstores/milvus.py:1095]: ditto

File: libs/partners/milvus/langchain_milvus/vectorstores/milvus.py [hunk 6]
```diff
@@ -1170,7 +1195,7 @@ def delete(  # type: ignore[no-untyped-def]
     def from_texts(
         cls,
         texts: List[str],
-        embedding: Embeddings,
+        embedding: Union[Embeddings, BaseSparseEmbedding],  # type: ignore
```
Comment (zc277584121) [inline libs/partners/milvus/langchain_milvus/vectorstores/milvus.py:1198]: ditto

File: libs/partners/milvus/langchain_milvus/vectorstores/milvus.py (hunk 3 above)
Comment (ohadeytan) [reply libs/partners/milvus/langchain_milvus/vectorstores/milvus.py:840]: done

File: libs/partners/milvus/langchain_milvus/vectorstores/milvus.py (hunk 4 above)
Comment (ohadeytan) [reply libs/partners/milvus/langchain_milvus/vectorstores/milvus.py:1001]: done

File: libs/partners/milvus/langchain_milvus/vectorstores/milvus.py (hunk 5 above)
Comment (ohadeytan) [reply libs/partners/milvus/langchain_milvus/vectorstores/milvus.py:1093]: done

File: libs/partners/milvus/langchain_milvus/vectorstores/milvus.py (hunk 6 above)
Comment (ohadeytan) [reply libs/partners/milvus/langchain_milvus/vectorstores/milvus.py:1196]: done

**Comment by zc277584121** [comment]
gltm, but there are some lint issues, besides them all are ok to me

**Comment by ohadeytan** [comment]
Thank you @efriis and @zc277584121!

Can you share your thoughts about extending `Milvus` class further to support multi vector search (aka hybrid search)?
The main change would be to allow getting a list of embeddings as a parameter, and enable searching with different configurations.

**Comment by zc277584121** [comment]
@ohadeytan 
I think it may not be easy to perform hybrid search in VectorStore at present, and it may be more appropriate to implement it in the Retriever class.
milvus SDK currently has two layers of interfaces, one is [ORM](https://milvus.io/api-reference/pymilvus/v2.4.x/ORM/Collection/Collection.md), which is at the bottom layer, and the other is [milvusCilent](https://milvus.io/docs/quickstart.md) , which is closer to users and is the mainstream in the future.
https://python.langchain.com/v0.2/docs/integrations/retrievers/milvus_hybrid_search/
This is MilvusCollectionHybridSearchRetriever implemented using the milvus ORM interface. Our milvusCilent has not yet implemented hybrid search. After milvusCilent supports it, we plan to mainly recommend MilvusHybridSearchRetriever  in the future, which is supported by milvusCilent, instead of Milvus**Collection**HybridSearchRetriever.
However, no matter ORM or milvusclient, they have to combined with the Retriever class, which means that we can only use milvus ORM or milvusclient to insert data, then wrap it with a retriever instance, and then use this retriever to retrieve. This is a little inconvenient. But this is the solution of other similar products we have seen so far.
Well, these are just designs, they may be overturned at any time. Any ideas and suggestions can be discussed. Appreciate any ideas and suggestions.

**Comment by LukaszCmielowski** [comment]
I would like to understand what are the next steps in regards to hybrid-search support in LangChain.
The comment https://github.com/langchain-ai/langchain/pull/25284#issuecomment-2320691026 seems to focus on retrieval capabilities; however the question is about collection creation (VectorStore ?). In the provided example:  https://python.langchain.com/v0.2/docs/integrations/retrievers/milvus_hybrid_search it is being done using `pymilvus` directly.

So, the question is - what are the next steps and who can recommend / decide on final shape/solution.

**Comment by zc277584121** [comment]
@LukaszCmielowski https://milvus.io/docs/milvus_hybrid_search_retriever.md Now that we have completed the next step, here is the latest recommended use of hybrid search, please refer to

**Pull request merged**